src/core/mempool.rs

```diff
@@ -386,52 +435,136 @@ impl MemPoolDB {
     {
-        // Want to consider transactions with
-        // height > max(-1, tip_height - (MEMPOOL_MAX_TRANSACTION_AGE + 1))
-        let min_height = tip_height.checked_sub(MEMPOOL_MAX_TRANSACTION_AGE + 1);
-        let mut curr_page = 0;
-
-        let mut next_nonce =
-            match MemPoolDB::get_next_nonce(&self.db, min_height, tip_height, None)? {
-                None => {
-                    return Ok(());
-                }
-                Some(nonce) => nonce,
-            };
+        let min_height = (tip_height as i64)
+            .checked_sub((MEMPOOL_MAX_TRANSACTION_AGE + 1) as i64)
+            .unwrap_or(-1);
+        let max_height = tip_height as i64;
+
+        let page_size = 1000;
+        let mut offset = 0;
+
+        let min_cumulative_fees = settings.min_cumulative_fee;
+        let min_tx_fee = settings.min_tx_fee;
+
+        let mut done = false;
+        let deadline = get_epoch_time_ms() + (settings.max_walk_time_ms as u128);
+        let mut total_considered = 0;
+        let mut total_origins = 0;
 
-        loop {
-            let available_txs = MemPoolDB::get_txs_at_nonce_and_offset(
-                &self.db, min_height, tip_height, next_nonce, curr_page,
+        test_debug!(
+            "Mempool walk for {}ms, min tx fee {}, min cumulative tx fees {}",
+            settings.max_walk_time_ms,
+            min_tx_fee,
+            min_cumulative_fees
+        );
+
+        while !done {
+            if deadline <= get_epoch_time_ms() {
+                test_debug!("Mempool iteration deadline exceeded");
+                break;
+            }
+
+            let origin_addresses = self.find_origin_addresses_by_descending_fees(
+                min_height,
+                max_height,
+                min_cumulative_fees,
+                offset * page_size,
+                page_size,
             )?;
             debug!(
-                "Have {} transactions at nonce={}",
-                available_txs.len(),
-                next_nonce,
+                "Consider {} origin addresses between {},{} with min_fee {}, min_cumulative_fee {}",
+                origin_addresses.len(),
+                min_height,
+                max_height,
+                min_tx_fee,
+                min_cumulative_fees
             );
 
-            if available_txs.len() > 0 {
-                todo(available_txs)?;
-                curr_page += 1;
-            } else {
-                curr_page = 0;
-                next_nonce = match MemPoolDB::get_next_nonce(
-                    &self.db,
-                    min_height,
-                    tip_height,
-                    Some(next_nonce),
-                )? {
-                    None => {
-                        return Ok(());
+            if origin_addresses.len() == 0 {
+                debug!("No more origin addresses to consider");
+                break;
+            }
+
+            for origin_address in origin_addresses.iter() {
```

Comment:
It seems to me like what you are trying to do here is:
1) get the origin addresses that are trying to spend the most
2) for those addresses, find the lowest nonces.

There is something to be said for the "simpler" strategy that you just sort by transaction fee.

For exmaple, to "cancel" a transaction in ETH, you submit a "0 transfer" transaction, with the same nonce, and a higher transaction fee than the transaction you want to cancel.

https://info.etherscan.com/how-to-cancel-ethereum-pending-transactions/

Maybe you can 1) sort by transaction fee, and then, to break ties 2) sort by nonce. This would support chaining.

But, if there is no tie on transaction fee, then don't involve the nonces, to allow transaction cancelation.

Users should have the awareness on their side to not submit the chained transaction (nonce = n + k) with a higher fee than the base transaction that is still to process (nonce = n).

Reply:
None of the Hiro tooling currently supports RBF (ethereum's "cancel" operation), so if the logic only ordered by highest fee, your transactions could get stuck indefinitely under mempool congestion unless you're willing to manually change the fee of your transaction and re-sign and re-submit it (something only developers can be expected to do, and even they wouldn't be happy about it).  In this query, because the cumulative transaction fee for an origin is what gets used to prioritize transactions, you can use the current tooling to un-stuck your transactions simply by sending more of them.

Reply:
I think that using the raw fee of transactions, rather than the cumulative fee of chained transactions would be a better approach here. With the way it is currently written, this mining logic will _strongly_ favor the chattiest transaction submitters, such that if blocks are all full, you'd expect that any origin that only submits one transaction at a time will almost always get crowded out. I know that most tooling doesn't support RBF yet, but that is an easier change for wallets, libraries, etc. to roll out (and a feature that they'd want to support anyways), rather than miner block assembly, so I'd definitely advocate for using high fees.

Reply:
Yup; changed to prioritizing by just tx fee

Reply:
Can you change the iteration to no longer consider all of an origin's transaction as a block? This new ordering still would end up favoring chatty submitters (if I have a single high fee transaction, then all of my transactions would be considered for inclusion before considering anyone else's transactions).

Reply:
The downside to this is that it will lead to starvation if there are few distinct origin addresses (i.e. when the mempool isn't congested).  Maybe I can fix this too -- now that `iterate_candidates()` takes a `ClarityTx`, the miner ought to be able to do multiple mempool passes when filling a block, thereby mining multiple transactions from the same origin but only after all other origins have been considered.

Reply:
I think that's fine -- I'm more worried about the current behavior where a single origin address could monopolize the miner's attempts just by having one transaction be the highest fee tx.

Reply:
I made it so that all origin addresses that clear the minimum fee will be considered in a round-robin fashion, and that if time and space permits, the miner will continue to cycle through origin addresses until the block is filled.  I've added an integration test that verifies that multiple transactions from the same origin will be included in a block, but *only after* all origins have been considered first.  Sound good?

Reply:
Perfect!